virttest/utils_netperf.py

```diff
@@ -126,9 +126,10 @@ def pack_compile(self, compile_option=""):
         build_type = {"aarch64": "aarch64-unknown-linux-gnu"}
         build_arch = self.session.cmd_output("arch", timeout=60)
         np_build = build_type.get(build_arch, build_arch)
-        setup_cmd = "./configure --build=%s %s > /dev/null" % (np_build,
-                                                               compile_option)
-        setup_cmd += " && make > /dev/null"
+        setup_cmd = "./configure --build=%s %s " \
+                    + "> /tmp/netperf_compile.log 2>&1" \
+                    % (np_build, compile_option)
+        setup_cmd += " && make >> /tmp/netperf_compile.log 2>&1"
```

Comment:
1. this change is different from the commit msg,
> Redirect the output of configure and make to /dev/null

but here we redirect the output of configure and make to a file.

2. if we really need to redirect to a file, it is better to place it in logdir of the test.
and make the file path a param and default it to logdir.

Reply:
Redirect to /dev/null now. Thanks.

Reply:
When a shell command lasts long and outputs to stdout&stderr, "get exit status will fail".

python3.4/site-packages/aexpect/client.py:
`     def cmd_status_output(self, cmd, timeout=60, internal_timeout=None,
                           print_func=None, safe=False):
         """
         Send a command and return its exit status and output.
 
         :param cmd: Command to send (must not contain newline characters)
         :param timeout: The duration (in seconds) to wait for the prompt to
                 return
         :param internal_timeout: The timeout to pass to read_nonblocking
         :param print_func: A function to be used to print the data being read
                 (should take a string parameter)
         :param safe: Whether using safe mode when execute cmd.
                 In serial sessions, frequently the kernel might print debug or
                 error messages that make read_up_to_prompt to timeout. Let's try
                 to be a little more robust and send a carriage return, to see if
                 we can get to the prompt when safe=True.
 
         :return: A tuple (status, output) where status is the exit status and
                 output is the output of cmd
         :raise ShellTimeoutError: Raised if timeout expires
         :raise ShellProcessTerminatedError: Raised if the shell process
                 terminates while waiting for output
         :raise ShellStatusError: Raised if the exit status cannot be obtained
         :raise ShellError: Raised if an unknown error occurs
         """
         o = self.cmd_output(cmd, timeout, internal_timeout, print_func, safe)
         try:
             # Send the 'echo $?' (or equivalent) command to get the exit status
             s = self.cmd_output(self.status_test_command, 10, internal_timeout,
                                 print_func, safe)
         except ShellError:
             raise ShellStatusError(cmd, o)
 
         # Get the first line consisting of digits only
         digit_lines = [l for l in s.splitlines() if l.strip().isdigit()]
         if digit_lines:
             return int(digit_lines[0].strip()), o
         else:
             raise ShellStatusError(cmd, o)
`

Example:
`2019-01-08 05:30:00,785 stacktrace       L0044 ERROR| Reproduced traceback from: /var/tmp/libvirt-ci/runtest/avocado-vt/avocado-vt/avocado_vt/test.py:438
2019-01-08 05:30:00,829 stacktrace       L0047 ERROR| Traceback (most recent call last):
2019-01-08 05:30:00,829 stacktrace       L0047 ERROR|   File "/var/tmp/libvirt-ci/runtest/avocado-vt/avocado-vt/virttest/utils_netperf.py", line 130, in pack_compile
2019-01-08 05:30:00,829 stacktrace       L0047 ERROR|     self.session.cmd(cmd, timeout=1200)
2019-01-08 05:30:00,829 stacktrace       L0047 ERROR|   File "/root/.libvirt-ci-venv-ci-runtest-wyFMe0/lib/python3.6/site-packages/aexpect/client.py", line 1286, in cmd
2019-01-08 05:30:00,829 stacktrace       L0047 ERROR|     print_func)
2019-01-08 05:30:00,829 stacktrace       L0047 ERROR|   File "/root/.libvirt-ci-venv-ci-runtest-wyFMe0/lib/python3.6/site-packages/aexpect/client.py", line 1226, in cmd_status_output
2019-01-08 05:30:00,829 stacktrace       L0047 ERROR|     raise ShellStatusError(cmd, o)
2019-01-08 05:30:00,829 stacktrace       L0047 ERROR| aexpect.exceptions.ShellStatusError: Could not get exit status of command: 'cd /var/tmp/  && tar jxvf netperf-2.6.0.tar.bz2 && cd /var/tmp/netperf-2.6.0  && ./configure --build=$(arch)  > /dev/null  && make > /dev/null '    (output: 'netperf-2.6.0/\nnetperf-2.6.0/README.windows\nnetperf-2.6.0/config.guess\nnetperf-2.6.0/netperf.spec.in\nnetperf-2.6.0/Makefile.in\nnetperf-2.6.0/COPYING\nnetperf-2.6.0/Makefile.am\nnetperf-2.6.0/configure.ac\nnetperf-2.6.0/autogen.sh\nnetperf-2.6.0/doc/\nnetperf-2.6.0/doc/netserver.man\nnetperf-2.6.0/doc/netperf.texi\nnetperf-2.6.0/doc/Makefile.in\nnetperf-2.6.0/doc/Makefile.am\nnetperf-2.6.0/doc/examples/\nnetperf-2.6.0/doc/examples/Makefile.in\nnetperf-2.6.0/doc/examples/Makefile.am\nnetperf-2.6.0/doc/examples/tcp_rr_script\nnetperf-2.6.0/doc/examples/packet_byte_script\nnetperf-2.6.0/doc/examples/tcp_stream_script\nnetperf-2.6.0/doc/examples/tcp_range_script\nnetperf-2.6.0/doc/examples/udp_rr_script\nnetperf-2.6.0/doc/examples/arr_script\nnetperf-2.6.0/doc/examples/sctp_stream_script\nnetperf-2.6.0/doc/examples/snapshot_script\nnetperf-2.6.0/doc/examples/udp_stream_script\nnetperf-2.6.0/doc/netperf.man\nnetperf-2.6.0/doc/texinfo.tex\nnetperf-2.6.0/doc/netperf.info\nnetperf-2.6.0/doc/netperf_old.ps\nnetperf-2.6.0/README.solaris\nnetperf-2.6.0/Release_Notes\nnetperf-2.6.0/README.osx\nnetperf-2.6.0/mkinstalldirs\nnetperf-2.6.0/configure\nnetperf-2.6.0/missing\nnetperf-2.6.0/src/\nnetperf-2.6.0/src/netsys_solaris.c\nnetperf-2.6.0/src/netrt_rtnetlink.c\nnetperf-2.6.0/src/netserver.c\nnetperf-2.6.0/src/netdrv_none.c\nnetperf-2.6.0/src/NetServerDir/\nnetperf-2.6.0/src/NetServerDir/sources\nnetperf-2.6.0/src/NetServerDir/makefile\nnetperf-2.6.0/src/NetServerDir/inet_ntop.c\nnetperf-2.6.0/src/Makefile.in\nnetperf-2.6.0/src/netdrv_ethtool.c\nnetperf-2.6.0/src/netcpu_perfstat.c\nnetperf-2.6.0/src/Makefile.am\nnetperf-2.6.0/src/netslot_none.c\nnetperf-2.6.0/src/nettest_sdp.c\nnetperf-2.6.0/src/nettest_unix.c\nnetperf-2.6.0/src/netslot_linux.c\nnetperf-2.6.0/src/netsys_linux.c\nnetperf-2.6.0/src/Makefile.uw\nnetperf-2.6.0/src/nettest_dlpi.c\nnetperf-2.6.0/src/netrt_none.c\nnetperf-2.6.0/src/netcpu_procstat.c\nnetperf-2.6.0/src/netcpu_looper.c\nnetperf-2.6.0/src/netcpu_kstat.c\nnetperf-2.6.0/src/netcpu_sysctl.c\nnetperf-2.6.0/src/netslot_solaris.c\nnetperf-2.6.0/src/netrt_rtmget.c\nnetperf-2.6.0/src/netsys_none.c\nnetperf-2.6.0/src/nettest_sctp.c\nnetperf-2.6.0/src/netlib.h\nnetperf-2.6.0/src/netdrv_solaris.c\nnetperf-2.6.0/src/nettest_sdp.h\nnetperf-2.6.0/src/netcpu_pstat.c\nnetperf-2.6.0/src/nettest_xti.h\nnetperf-2.6.0/src/netcpu_kstat10.c\nnetperf-2.6.0/src/netsh.c\nnetperf-2.6.0/src/net_uuid.c\nnetperf-2.6.0/src/netperf_version.h.in\nnetperf-2.6.0/src/missing/\nnetperf-2.6.0/src/missing/Makefile.in\nnetperf-2.6.0/src/missing/Makefile.am\nnetperf-2.6.0/src/missing/m4/\nnetperf-2.6.0/src/missing/m4/Makefile.in\nnetperf-2.6.0/src/missing/m4/Makefile.am\nnetperf-2.6.0/src/missing/m4/herrno.m4\nnetperf-2.6.0/src/missing/m4/socklent.m4\nnetperf-2.6.0/src/missing/m4/sockaddrin6.m4\nnetperf-2.6.0/src/missing/m4/type_socklen_t.m4\nnetperf-2.6.0/src/missing/m4/in6addr.m4\nnetperf-2.6.0/src/missing/m4/sockinttypes.m4\nnetperf-2.6.0/src/missing/m4/salen.m4\nnetperf-2.6.0/src/missing/getaddrinfo.h\nnetperf-2.6.0/src/missing/inet_ntop.c\nnetperf-2.6.0/src/missing/getaddrinfo.c\nnetperf-2.6.0/src/nettest_bsd.h\nnetperf-2.6.0/src/netcpu_none.c\nnetperf-2.6.0/src/netsh.h\nnetperf-2.6.0/src/netperf.c\nnetperf-2.6.0/src/netcpu_osx.c\nnetperf-2.6.0/src/NetPerfDir/\nnetperf-2.6.0/src/NetPerfDir/sources\nnetperf-2.6.0/src/NetPerfDir/makefile\nnetperf-2.6.0/src/NetPerfDir/inet_ntop.c\nnetperf-2.6.0/src/netcpu_ntperf.c\nnetperf-2.6.0/src/netcpu.h\nnetperf-2.6.0/src/dscp.c\nnetperf-2.6.0/src/nettest_xti.c\nnetperf-2.6.0/src/netcpu_pstatnew.c\nnetperf-2.6.0/src/nettest_bsd.c\nnetperf-2.6.0/src/nettest_omni.c\nnetperf-2.6.0/src/netsec_linux.c\nnetperf-2.6.0/src/netsys_hpux11i.c\nnetperf-2.6.0/src/nettest_sctp.h\nnetperf-2.6.0/src/nettest_dlpi.h\nnetperf-2.6.0/src/netslot_ux1131.c\nnetperf-2.6.0/src/netsec_none.c\nnetperf-2.6.0/src/nettest_unix.h\nnetperf-2.6.0/src/netlib.c\nnetperf-2.6.0/src/netperf_version.h\nnetperf-2.6.0/src/dirs\nnetperf-2.6.0/src/hist.h\nnetperf-2.6.0/NEWS\nnetperf-2.6.0/depcomp\nnetperf-2.6.0/install-sh\nnetperf-2.6.0/m4/\nnetperf-2.6.0/m4/.svn/\nnetperf-2.6.0/m4/.svn/tmp/\nnetperf-2.6.0/m4/.svn/tmp/text-base/\nnetperf-2.6.0/m4/.svn/tmp/prop-base/\nnetperf-2.6.0/m4/.svn/tmp/props/\nnetperf-2.6.0/m4/.svn/text-base/\nnetperf-2.6.0/m4/.svn/prop-base/\nnetperf-2.6.0/m4/.svn/entries\nnetperf-2.6.0/m4/.svn/all-wcprops\nnetperf-2.6.0/m4/.svn/props/\nnetperf-2.6.0/m4/m4/\nnetperf-2.6.0/m4/m4/.svn/\nnetperf-2.6.0/m4/m4/.svn/tmp/\nnetperf-2.6.0/m4/m4/.svn/tmp/text-base/\nnetperf-2.6.0/m4/m4/.svn/tmp/prop-base/\nnetperf-2.6.0/m4/m4/.svn/tmp/props/\nnetperf-2.6.0/m4/m4/.svn/text-base/\nnetperf-2.6.0/m4/m4/.svn/text-base/salen.m4.svn-base\nnetperf-2.6.0/m4/m4/.svn/text-base/sockaddrin6.m4.svn-base\nnetperf-2.6.0/m4/m4/.svn/text-base/sockinttypes.m4.svn-base\nnetperf-2.6.0/m4/m4/.svn/prop-base/\nnetperf-2.6.0/m4/m4/.svn/entries\nnetperf-2.6.0/m4/m4/.svn/all-wcprops\nnetperf-2.6.0/m4/m4/.svn/props/\nnetperf-2.6.0/m4/m4/sockaddrin6.m4\nnetperf-2.6.0/m4/m4/sockinttypes.m4\nnetperf-2.6.0/m4/m4/salen.m4\nnetperf-2.6.0/README.ovms\nnetperf-2.6.0/README.hpux\nnetperf-2.6.0/INSTALL\nnetperf-2.6.0/aclocal.m4\nnetperf-2.6.0/config.sub\nnetperf-2.6.0/AUTHORS\nnetperf-2.6.0/README\nnetperf-2.6.0/README.vmware\nnetperf-2.6.0/ChangeLog\nnetperf-2.6.0/inet_ntop.c\nnetperf-2.6.0/config.h.in\nnetperf-2.6.0/README.aix\n')`

Reply:
In actual code we have,
https://github.com/avocado-framework/avocado-vt/blob/a06a2447864a16f24ed747ec4e80cdcf5dcb8f26/virttest/utils_netperf.py#L129-L137

where `self.session.cmd` will only return output of the command,
1. if you need the exit status and use it for assertion, then we have to use `self.session.cmd_status`.
2. we can increase the `timeout` parameter if the command takes more time.

The current changes would append stderr with stdout and redirect it to /dev/null, so it will not fix the problem.

Reply:
In function cmd(), it calls cmd_status_output() which returns both exit status and output.
Either stderr or stdout will cause "exit status" check fail, so we need to redirect both stderr and stdout to /dev/null.

Reply:
okay, could you able to test the fix ?

Reply:
`# avocado run --vt-type libvirt --vt-machine-type q35 virsh.migrate_vm.positive_testing.live_migration.stress.host.netperf_network.with_postcopy --vt-connect-uri qemu:///system
JOB ID     : 736280338679f4aa6c17ed6ae36dd186e83f25f7
JOB LOG    : /root/avocado/job-results/job-2019-01-21T15.27-7362803/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.migrate_vm.positive_testing.live_migration.stress.host.netperf_network.with_postcopy: PASS (196.68 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 233.85 s
`